libraries/rush-lib/src/logic/operations/DependencyAnalysisPlugin.ts

```diff
@@ -0,0 +1,330 @@
+        });
+        const projectStartTime: number = Math.max(firstEndTime, ...dependencyBuildTimings);
+        const accumulativeBuildTime: number = projectStartTime + (buildTimes.get(project.project) ?? -1);
+        accumulativeBuildTimeMap.set(project.project, Number(accumulativeBuildTime.toFixed(2)));
+      }
+    }
+  }
+  const totalBuildTime: number = Math.max(...Array.from(accumulativeBuildTimeMap.values()));
+  return totalBuildTime;
+}
+
+export function simulateBuildTime(filename: string, terminal: ITerminal): void {
+  const telemetryFile: ITelemetryData = JsonFile.load(filename)[0];
+  const buildTimes: IBuildTimeRecord[] = telemetryFile.buildTimings ?? [];
+  const dependencyGraph: IDependencyGraph = telemetryFile.dependencyGraph ?? {};
+
+  const simulatedBuildTimes: number[] = [];
+  const buildTimingMap: Map<string, number> = new Map();
+
+  buildTimes.forEach((project: IBuildTimeRecord) => {
+    buildTimingMap.set(project.project, project.buildTime);
+  });
+
+  for (let numCores: number = 1; numCores < MAX_CORES; numCores++) {
+    const simulatedBuildTime: number = calculateBuildTime(dependencyGraph, buildTimingMap, numCores);
+    if (simulatedBuildTimes.length === 0) {
+      simulatedBuildTimes.push(simulatedBuildTime);
+    } else {
+      if (simulatedBuildTimes[simulatedBuildTimes.length - 1] * 0.999 > simulatedBuildTime) {
+        simulatedBuildTimes.push(simulatedBuildTime);
+      } else {
+        simulatedBuildTimes.push(simulatedBuildTime);
+        writeSimulationSummary(terminal, simulatedBuildTimes);
+        return;
+      }
+    }
+  }
+}
+function writeSimulationSummary(terminal: ITerminal, simulatedBuildTimes: number[]): void {
+  terminal.writeLine(`Simulating build time given number of cores:`);
+
+  simulatedBuildTimes.forEach((value, index) => {
+    const padding: string = ' '.repeat(20 - (index + 1).toString().length);
+    terminal.writeLine(`  Number of cores: ${index + 1}${padding} ${value} seconds`);
+  });
+
+  terminal.writeLine(
+    `The optimal number of cores to build your project is ${simulatedBuildTimes.length - 1}`
+  );
+}
+
+function writeSummary(terminal: ITerminal, buildTimeList: IBuildTimeSavedRecord[]): void {
+  let longestTaskName: number = 0;
+  for (const project of buildTimeList) {
+    const nameLength: number = (project.project || '').length;
+    if (nameLength > longestTaskName) {
+      longestTaskName = nameLength;
+    }
+  }
+
+  terminal.writeLine(
+    `If the following project had a 0 second build time, then the total build time will change by:`
+  );
+
+  for (const project of buildTimeList) {
+    const padding: string = ' '.repeat(longestTaskName - (project.project || '').length);
+    terminal.writeLine(`  ${project.project}${padding}    ${project.secondsSaved} seconds saved`);
+  }
+}
+
+export function _printDependencyAnalysis(terminal: ITerminal, result: IExecutionResult): void {
+  const dependencyGraph: IDependencyGraph = makeDependencyGraph();
```

Comment:
The execution-time dependency graph is directly available by examining `result.operationResults.keys()`; the `Operation` instances all list their dependencies on `.dependencies` and consumers on `.consumers`.

Reply:
This refactor makes the code simpler and cleaner. Thank you for pointing me to `result.operationResults`